autonomy/chain/base.py

```diff
@@ -0,0 +1,110 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2022 Valory AG
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Chain interaction base classes."""
+
+import json
+from enum import Enum
+from typing import Any, Dict, Iterable, List, Optional
+
+from aea.crypto.base import Crypto, LedgerApi
+
+from autonomy.chain.config import ContractConfig
+from autonomy.data import DATA_DIR
+
+
+ABI_DIR = DATA_DIR / "abis"
+
+ContractInterfaceType = Any
+
+
+def get_abi(filename: str) -> Dict:
+    """Service contract ABI."""
+    with (ABI_DIR / filename).open(mode="r", encoding="utf-8") as fp:
+        return json.load(fp=fp)
+
+
+class BaseContract:  # pylint: disable=too-few-public-methods
```

Comment:
We have similar classes in open-aea. Duplication is not ideal

Reply:
Yes we do have a class on open-aea but it's a component implementation not a stand alone contract interface implementation. If we choose to use `aea.contracts.base.Contract` it will make things a bit more complicated than they need to be. To being with to initialise `aea.contracts.base.Contract` we need a `aea.configurations.base.ContractConfig` object, which is a bit out of context to use here since to interact with the registry contracts at the framework level we only need the ledger api and contract address whereas `aea.configurations.base.ContractConfig` contains much more information then it's required.

What I'm trying to say is let's not try to over generalise this implementation or it'll just make things more complicated and bloated